text/0000-PXE-Booting-UEFI.md

```diff
@@ -0,0 +1,228 @@
+system model classes.
+
+Initial testing was done to get EFI working on PXELinux and it looks like
+stable version of Syslinux (= PXELinux), which is currently 6.02, does not
+appear to support PXE booting correctly (at least not with TianoCore EFI ROM in
+QEMU/KVM). The only version which was found working was 6.03-pre20.
+
+Additionally, there is no PXELinux version with EFI support in RHEL6 and RHEL7
+and both systems ship with Grub (versions 1 and 2 respectively). Grub is more
+mature when it comes to EFI support and since it can boot via PXE and Foreman
+already supports Grub1, it is more feasible to get EFI working via officially
+supported Grub. To have complete support matrix, PXELinux will be part of this
+implementation, but hardware support (and thus testing) will be limited.
+
+The first idea was to simply deploy both BIOS and UEFI configurations during
+orchestration and configure hosts to pick up the correct loader. This is
+possible in ISC DHCPD:
+
+```
+option space PXE;
+option arch code 93 = unsigned integer 16; # RFC4578
+
+subnet 10.0.0.0 netmask 255.255.255.0 {
+    option routers 10.0.0.254;
+    range 10.0.0.2 10.0.0.253;
+
+    class "pxeclients" {
+        match if substring (option vendor-class-identifier, 0, 9) = "PXEClient";
+        next-server 10.0.0.1;
+        if option arch = 00:06 {
+            filename "grub2/bootia32.efi";
+        } else if option arch = 00:07 {
+            filename "grub2/bootx64.efi";
+        } else {
+            filename "pxelinux.0";
+        }
+    }
+
+    host example-ia32 {
+        hardware ethernet XX:YY:ZZ:11:22:33;
+        fixed-address 10.0.0.2;
+    }
+}
+```
+
+The major issue with this approach is hosts in Foreman both in boot and
+production mode have their own DHCP reservation which does not provide a syntax
+to do if statement. Option "filename" must be a constant. For this reason both
+DHCP filename and TFTP configuration must be deployed correctly for each
+individual host.
+
+Both Grub 1 and 2 have MAC-based config file retrieval built-in and both have
+different configuration syntax, therefore new PXE kind (Grub2) must be added.
+Currently supported PXE combinations are:
+
+Variant  | Template kind | Arch | DHCP filename | MAC config | Default config
+--- | --- | --- | --- | --- | ---
+SYSLINUX | PXELinux |	BIOS | pxelinux.0 |	pxelinux.cfg/01-aa-bb-cc-dd-ee-ff |	pxelinux.cfg/default
+PXEGRUB |	PXEGrub | BIOS | Solaris-* | boot/grub/menu.lst.01aabbccddeeff | boot/grub/menu.lst
+POAP | POAP |	- |	poap.cfg/aabbccddeeff | poap.cfg/AABBCCDDEEFF | -
+ZTP | ZTP | - | poap.cfg/aabbccddeeff | poap.cfg/AABBCCDDEEFF | -
+
+PXELinux covers all x86 systems, PXEGrub is used for Solaris (firmware filename
+is Solaris-*-pxegrub) and POAP/ZTP have built-in loaders and only downloads
+configuration files. This feature aims for this support matrix:
+
+Variant |	Template kind | Arch | DHCP filename(s) | MAC config | Default config | Comment
+--- | --- | --- | --- | --- | --- | ---
```

Comment:
Github is not very good in tables, you need to use horizontal scrollbar to see the last column, can't fix as the table has lot of data in it.